gufe/storage/utils.py

```diff
@@ -0,0 +1,85 @@
+from typing import NamedTuple, Dict, Tuple, Any
+
+import hashlib
+
+import importlib
+import inspect
+
+
+REMAPPED_CLASSES: Dict[Tuple[str, str], Tuple[str, str]] = {}
+"""Mapping of old module/class names to new ones. Backward compatibility.
+
+This should be a mapping of the Tuple[str, str] giving (old_module,
+old_class) to Tuple[str, str] giving (new_module, new_class)
+"""
+
+OBJECT_REGISTRY: Dict[str, Any] = {}
+"""Mapping of pathlike location to object for objects we need to cache.
+
+This prevents us from making extraneous copies of objects that appear at
+multiple places in the data structure.
+"""
+
+
+class SerializationInfo(NamedTuple):
+    """
+    Parameters
+    ----------
+    bytes_data: byes
+        bytes to be stored to disk for this object
+    metadata: Dict
+        metadata about this object, contains the following fields:
+
+        * ``:path:``: The storage path for this object
+        * ``:class:``: The qualname for this class of this object
+        * ``:module:``: The module where the class is found
+        * ``:md5:``: The md5 hexdigest for this object. In order to preserve
+          this value across code versions, this hash is based on the
+          arguments with non-default values, so it does not necessarily
+          correspond to the md5 of the associated ``bytes_data``.
+    """
+    bytes_data: bytes
+    metadata: Dict
+
+    @property
+    def path(self):
+        return self.metadata[':path:']
+
+    @property
+    def md5(self):
+        return self.metadata[':md5:']
+
+    @staticmethod
+    def create_metadata(obj, bytes_data, path_pattern):
+        md5 = hashlib.md5(bytes_data).hexdigest()
+        cls = obj.__class__
+        return {
+            ":path:": path_pattern.format(md5=md5[:10]),
```

Comment:
Agreed -- this should be the `base64.b32encode` or `base64.b32hexencode` of the full md5. (Using b32 so we don't need to worry about case-dependent labels).

I did this as is simply because I didn't want full hex versions; figured if GitHub gets by on 7 hex digits, 10 would probably be enough.